tracee-rules/engine/engine.go

```diff
@@ -62,24 +67,34 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 	return &engine
 }
 
+// signatureStartFunc is the signature handling business logics.
+func (engine *Engine) signatureStartFunc(signature types.Signature, c chan types.Event) {
+	for {
+		select {
+		case e, ok := <-c:
+			if !ok {
+				// channel closed
+				return
+			}
+			if err := signature.OnEvent(e); err != nil {
+				meta, _ := signature.GetMetadata()
+				log.Printf("error handling event by signature %s: %v", meta.Name, err)
+			}
+		}
+	}
+}
+
 // Start starts processing events and detecting signatures
 // it runs continuously until stopped by the done channel
 // once done, it cleans all internal resources, which means the engine is not reusable
 // note that the input and output channels are created by the consumer and therefore are not closed
 func (engine *Engine) Start(done chan bool) {
 	go engine.consumeSources(done)
+	engine.signaturesMutex.RLock()
 	for s, c := range engine.signatures {
-		defer close(c)
```

Comment:
you removed the sig-channel close, and only added it in the signature unload. doesn't it mean we will leak channels if the engine stops? (unless someone unloads all the signatures manually which is unlikely)

Reply:
another option is to defer-unload all signatures in start

Reply:
also, see https://github.com/aquasecurity/tracee/pull/799

Reply:
ConsumeSources is a blocking call. 
I think UnloadSignature is should be the one to close the signature channel. deferring unload to all signatures can be an idea.